File: App/Windows/BeosztasVarazslo/Services/ScheduleCalculator.cs
using BeosztasVarazslo.Models;

namespace BeosztasVarazslo.Services;

public enum CellKind { Work, Vacation, Absence, Rest, Empty, NonWork }

public record CellInfo(double Hours, CellKind Kind, string? Code = null);

public class MonthSummary
{
    public double BaseHours { get; init; }
    public double EmploymentFactor { get; init; }
    public double RequiredFull { get; init; }
    public int VacationDays { get; init; }
    public int AbsenceDays { get; init; }
    public int WorkDays { get; init; }
    public double VacationHours { get; init; }
    public double AbsenceHours { get; init; }
    public double CarryIn { get; init; }
    public double EffectiveRequired { get; init; }
    public double ActualHours { get; init; }
    public double Balance { get; init; }
}

/// <summary>Óraszám-számítási motor: kötelező havi óraszám, ledolgozott óra, maradvány.</summary>
public static class ScheduleCalculator
{
    public static int DaysInMonth(int year, int month) => DateTime.DaysInMonth(year, month);

    public static bool IsWeekend(int year, int month, int day)
    {
        var dow = new DateTime(year, month, day).DayOfWeek;
        return dow is DayOfWeek.Saturday or DayOfWeek.Sunday;
    }

    public static string IsoDate(int year, int month, int day) => $"{year:D4}-{month:D2}-{day:D2}";

    public static bool IsHoliday(IReadOnlyDictionary<string, string> holidayMap, int year, int month, int day) =>
        holidayMap.ContainsKey(IsoDate(year, month, day));

    public static bool IsOfficeWorkday(IReadOnlyDictionary<string, string> holidayMap, int year, int month, int day) =>
        !IsWeekend(year, month, day) && !IsHoliday(holidayMap, year, month, day);

    public static CellInfo CellHours(string groupType, double groupDailyHours, List<ShiftType> shiftTypes, string? code, bool isOfficeWorkday)
    {
        if (groupType == GroupTypes.Office)
        {
            if (!isOfficeWorkday) return new CellInfo(0, CellKind.NonWork);
            return code switch
            {
                ShiftCodes.Vacation => new CellInfo(0, CellKind.Vacation),
                ShiftCodes.Absence or ShiftCodes.Sick => new CellInfo(0, CellKind.Absence),
                _ => new CellInfo(groupDailyHours, CellKind.Work, "M")
            };
        }

        if (string.IsNullOrEmpty(code)) return new CellInfo(0, CellKind.Empty);
        switch (code)
        {
            case ShiftCodes.Vacation: return new CellInfo(0, CellKind.Vacation);
            case ShiftCodes.Absence: case ShiftCodes.Sick: return new CellInfo(0, CellKind.Absence);
            case ShiftCodes.Rest: return new CellInfo(0, CellKind.Rest);
            default:
                var st = shiftTypes.Find(s => s.Code == code);
                return st != null ? new CellInfo(st.Hours, CellKind.Work, st.Code) : new CellInfo(0, CellKind.Empty);
        }
    }

    public static MonthSummary SummarizeMonth(
        string groupType, double groupDailyHours, List<ShiftType> shiftTypes,
        double employmentFactor, double baseMonthHours,
        IReadOnlyDictionary<int, string> codesByDay, IReadOnlyDictionary<string, string> holidayMap,
        double carryIn, int year, int month)
    {
        int dim = DaysInMonth(year, month);
        double actualHours = 0;
        int vacationDays = 0, absenceDays = 0, workDays = 0;

        for (int d = 1; d <= dim; d++)
        {
            codesByDay.TryGetValue(d, out var code);
            bool officeWorkday = IsOfficeWorkday(holidayMap, year, month, d);
            var info = CellHours(groupType, groupDailyHours, shiftTypes, code, officeWorkday);
            switch (info.Kind)
            {
                case CellKind.Work: actualHours += info.Hours; workDays++; break;
                case CellKind.Vacation: vacationDays++; break;
                case CellKind.Absence: absenceDays++; break;
            }
        }

        double requiredFull = baseMonthHours * employmentFactor;
        double vacationHours = vacationDays * groupDailyHours;
        double absenceHours = absenceDays * groupDailyHours;
        double effectiveRequired = requiredFull - vacationHours - absenceHours - carryIn;
        double balance = actualHours - effectiveRequired;

        return new MonthSummary
        {
            BaseHours = baseMonthHours, EmploymentFactor = employmentFactor, RequiredFull = requiredFull,
            VacationDays = vacationDays, AbsenceDays = absenceDays, WorkDays = workDays,
            VacationHours = vacationHours, AbsenceHours = absenceHours, CarryIn = carryIn,
            EffectiveRequired = effectiveRequired, ActualHours = actualHours, Balance = balance
        };
    }

    /// <summary>Egy adott napon, adott műszaktípus-kódonként hány dolgozó van beosztva.</summary>
    public static Dictionary<string, int> ShiftCoverage(List<ShiftType> shiftTypes, IEnumerable<string?> codesOfEmployeesThatDay)
    {
        var counts = shiftTypes.ToDictionary(s => s.Code, _ => 0);
        foreach (var code in codesOfEmployeesThatDay)
        {
            if (code != null && counts.ContainsKey(code)) counts[code]++;
        }
        return counts;
    }

    /// <summary>A cellára kattintva választható kódok (kód, megjelenített címke) párokban.</summary>
    public static List<(string Code, string Label)> CellOptions(string groupType, List<ShiftType> shiftTypes)
    {
        if (groupType == GroupTypes.Office)
        {
            return new List<(string, string)>
            {
                ("", "Munka"), (ShiftCodes.Vacation, "SZ – Szabadság"),
                (ShiftCodes.Sick, "BSZ – Beteg szabadság"), (ShiftCodes.Absence, "H – Hiányzás")
            };
        }
        var opts = new List<(string, string)> { ("", "—") };
        foreach (var st in shiftTypes) opts.Add((st.Code, $"{st.Code} – {st.Label}"));
        opts.Add((ShiftCodes.Vacation, "SZ – Szabadság"));
        opts.Add((ShiftCodes.Sick, "BSZ – Beteg szabadság"));
        opts.Add((ShiftCodes.Absence, "H – Hiányzás"));
        opts.Add((ShiftCodes.Rest, "P – Pihenőnap"));
        return opts;
    }
}

File: App/Android/app/src/main/java/hu/beosztasvarazslo/app/logic/ScheduleCalculator.kt
package hu.beosztasvarazslo.app.logic

import hu.beosztasvarazslo.app.data.ShiftTypeEntity
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.YearMonth

enum class CellKind { WORK, VACATION, ABSENCE, REST, EMPTY, NONWORK }

data class CellInfo(val hours: Double, val kind: CellKind, val code: String? = null)

data class MonthSummary(
    val baseHours: Double,
    val employmentFactor: Double,
    val requiredFull: Double,
    val vacationDays: Int,
    val absenceDays: Int,
    val workDays: Int,
    val vacationHours: Double,
    val absenceHours: Double,
    val carryIn: Double,
    val effectiveRequired: Double,
    val actualHours: Double,
    val balance: Double
)

const val GROUP_TYPE_OFFICE = "IRODA"
const val GROUP_TYPE_GENERAL = "ALTALANOS"

/** Óraszám-számítási motor: kötelező havi óraszám, ledolgozott óra, maradvány (a web app calc.js-ének Kotlin megfelelője). */
object ScheduleCalculator {
    const val CODE_VACATION = "SZ"
    const val CODE_SICK = "BSZ"
    const val CODE_ABSENCE = "H"
    const val CODE_REST = "P"

    fun daysInMonth(year: Int, month: Int): Int = YearMonth.of(year, month).lengthOfMonth()

    fun isWeekend(year: Int, month: Int, day: Int): Boolean {
        val dow = LocalDate.of(year, month, day).dayOfWeek
        return dow == DayOfWeek.SATURDAY || dow == DayOfWeek.SUNDAY
    }

    fun isoDate(year: Int, month: Int, day: Int): String = "%04d-%02d-%02d".format(year, month, day)

    fun isHoliday(holidayMap: Map<String, String>, year: Int, month: Int, day: Int): Boolean =
        holidayMap.containsKey(isoDate(year, month, day))

    fun isOfficeWorkday(holidayMap: Map<String, String>, year: Int, month: Int, day: Int): Boolean =
        !isWeekend(year, month, day) && !isHoliday(holidayMap, year, month, day)

    /** Egy nap (dolgozó, csoport) órahatása és jellege. */
    fun cellHours(
        groupType: String,
        groupDailyHours: Double,
        shiftTypes: List<ShiftTypeEntity>,
        code: String?,
        isOfficeWorkdayFlag: Boolean
    ): CellInfo {
        if (groupType == GROUP_TYPE_OFFICE) {
            if (!isOfficeWorkdayFlag) return CellInfo(0.0, CellKind.NONWORK)
            return when (code) {
                CODE_VACATION -> CellInfo(0.0, CellKind.VACATION)
                CODE_ABSENCE, CODE_SICK -> CellInfo(0.0, CellKind.ABSENCE)
                else -> CellInfo(groupDailyHours, CellKind.WORK, "M")
            }
        }
        if (code.isNullOrEmpty()) return CellInfo(0.0, CellKind.EMPTY)
        return when (code) {
            CODE_VACATION -> CellInfo(0.0, CellKind.VACATION)
            CODE_ABSENCE, CODE_SICK -> CellInfo(0.0, CellKind.ABSENCE)
            CODE_REST -> CellInfo(0.0, CellKind.REST)
            else -> {
                val st = shiftTypes.find { it.code == code }
                if (st != null) CellInfo(st.hours, CellKind.WORK, st.code) else CellInfo(0.0, CellKind.EMPTY)
            }
        }
    }

    /** Egy dolgozó adott havi összesítése (bejövő kézi órával csökkentve a kötelező óraszámot). */
    fun summarizeMonth(
        groupType: String,
        groupDailyHours: Double,
        shiftTypes: List<ShiftTypeEntity>,
        employmentFactor: Double,
        baseMonthHours: Double,
        codesByDay: Map<Int, String>,
        holidayMap: Map<String, String>,
        carryIn: Double,
        year: Int,
        month: Int
    ): MonthSummary {
        val dim = daysInMonth(year, month)
        var actualHours = 0.0
        var vacationDays = 0
        var absenceDays = 0
        var workDays = 0
        for (d in 1..dim) {
            val code = codesByDay[d]
            val officeWorkday = isOfficeWorkday(holidayMap, year, month, d)
            val info = cellHours(groupType, groupDailyHours, shiftTypes, code, officeWorkday)
            when (info.kind) {
                CellKind.WORK -> { actualHours += info.hours; workDays++ }
                CellKind.VACATION -> vacationDays++
                CellKind.ABSENCE -> absenceDays++
                else -> {}
            }
        }
        val requiredFull = baseMonthHours * employmentFactor
        val vacationHours = vacationDays * groupDailyHours
        val absenceHours = absenceDays * groupDailyHours
        val effectiveRequired = requiredFull - vacationHours - absenceHours - carryIn
        val balance = actualHours - effectiveRequired
        return MonthSummary(
            baseMonthHours, employmentFactor, requiredFull, vacationDays, absenceDays, workDays,
            vacationHours, absenceHours, carryIn, effectiveRequired, actualHours, balance
        )
    }

    /** Egy adott napon, adott műszaktípus-kódonként hány dolgozó van beosztva. */
    fun shiftCoverage(shiftTypes: List<ShiftTypeEntity>, codesOfEmployeesThatDay: Collection<String?>): Map<String, Int> {
        val counts = shiftTypes.associate { it.code to 0 }.toMutableMap()
        codesOfEmployeesThatDay.forEach { code ->
            if (code != null && counts.containsKey(code)) counts[code] = counts.getValue(code) + 1
        }
        return counts
    }

    /** A cellára kattintva választható kódok (kód, megjelenített címke) párokban. */
    fun cellOptions(groupType: String, shiftTypes: List<ShiftTypeEntity>): List<Pair<String, String>> {
        if (groupType == GROUP_TYPE_OFFICE) {
            return listOf(
                "" to "Munka", CODE_VACATION to "SZ – Szabadság",
                CODE_SICK to "BSZ – Beteg szabadság", CODE_ABSENCE to "H – Hiányzás"
            )
        }
        val opts = mutableListOf("" to "—")
        shiftTypes.forEach { opts.add(it.code to (it.code + " – " + it.label)) }
        opts.add(CODE_VACATION to "SZ – Szabadság")
        opts.add(CODE_SICK to "BSZ – Beteg szabadság")
        opts.add(CODE_ABSENCE to "H – Hiányzás")
        opts.add(CODE_REST to "P – Pihenőnap")
        return opts
    }
}
